src/System.ServiceModel.NetNamedPipe/src/System/ServiceModel/Channels/PipeConnectionInitiator.cs

```diff
@@ -0,0 +1,437 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.IO.MemoryMappedFiles;
+using System.IO.Pipes;
+using System.Runtime;
+using System.Runtime.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Principal;
+using System.ServiceModel.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class PipeConnectionInitiator : IConnectionInitiator
+    {
+        private readonly int _bufferSize;
+
+        public PipeConnectionInitiator(int bufferSize)
+        {
+            _bufferSize = bufferSize;
+        }
+
+        private Exception CreateConnectFailedException(Uri remoteUri, PipeException innerException)
+        {
+            return new CommunicationException(
+                SR.Format(SR.PipeConnectFailed, remoteUri.AbsoluteUri), innerException);
+        }
+
+        public async ValueTask<IConnection> ConnectAsync(Uri remoteUri, TimeSpan timeout)
```

Comment:
There are a lot of methods that pass around a TimeSpan timeout. Have you considered using a CancellationToken instead? That's the idiomatic way to do cancellation with Tasks.